sql/ui: surface network bytes sent in statements page

asubiotto opened the pull request:
Closes #55331 

This PR implements the conditional statistics collection and display described in phase 1 of the always-on EXPLAIN ANALYZE issue (#54556). The goal of this work is to provide a framework to surface more top-level statistics while always-on tracing is being worked on. The next steps here are to surface more top-level statistics we find interesting to the admin UI (e.g. query memory usage by @cathymw in #54340).

Once always-on tracing is completed, the code added to the conn executor to analyze trace data will be moved to `recordStatement`.

cc @jordanlewis

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55969)
<!-- Reviewable:end -->

asubiotto commented:
This is what the statement page looks like when statement diagnostics is enabled:
![image](https://user-images.githubusercontent.com/10560359/97179513-30bc1600-1799-11eb-809e-5caf748dab0b.png)

@dhartunian I'd appreciate if you could take a look at the `.tsx` change to see if I'm following best practices with the filtering of this information.

Another question I'd like to figure out is whether we want to think more about returning these stats in `StatementStatistics` grouped by node. What gives me pause here is how to handle when the number of nodes in the physical plan changes. What should we do then?

@RaduBerinde are you willing to be the primary reviewer on this? I think we'll be working pretty closely on observability stuff so it makes the most sense to me.

asubiotto commented:
cc @yuzefovich / @cathymw  on the first commit and fixing outbox bytes sent non-determinism.

yuzefovich reviewed (commented):
Nice work. I have a few nits.

Also, I'm curious why we're showing a decimal point in the number of "rows read" in the image, that looks not pretty, cc @dhartunian 

> Reviewed 13 of 13 files at r1, 5 of 5 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-:-MKbe1stB4lC9MLGuuUg:bj4cxqe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, @dhartunian, and @RaduBerinde)

___
*[pkg/sql/distsql_physical_planner.go, line 608 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MK_w_Zp9b6SusBMWwLn:-MK_w_Zp9b6SusBMWwLo:br3ee8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/distsql_physical_planner.go#L608)):*
> ```Go
> 	noEvalSubqueries bool
> 
> 	// If set, the flows for this physical plan will be passed to this function.
> ```

super nit: `s/this physical/the physical/` since it's not clear what "this" refers to.
___
*[pkg/sql/plan.go, line 338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbdOlEBaXoA9oJ_Ez5:-MKbdOlEBaXoA9oJ_Ez6:b4zucl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a366f2390d8d304a6d6ac3944e96e2e2d8b07bd/pkg/sql/plan.go#L338)):*
> ```Go
> 	// If we are collecting query diagnostics, flow diagrams are saved here.
> 	distSQLDiagrams []annotatedFlowDiagram
> 	// If we are collection query diagnostics, physical plans are saved here.
> ```

nit: `s/collection/collecting/`.
___
*[pkg/sql/traceanalyzer.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbSwo84ESkg9n9oEOI:-MKbSwo84ESkg9n9oEOJ:b3wv28k) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer.go#L22)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
> 	"github.com/gogo/protobuf/types"
> 	"github.com/pkg/errors"
> ```

nit: we should use our `cockroachdb/errors` :)
___
*[pkg/sql/traceanalyzer.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKb_gaE8aQVPkLv82VJ:-MKb_gaE8aQVPkLv82VK:bgxokxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer.go#L27)):*
> ```Go
> const nodeTagKey = "node"
> 
> type TraceAnalyzer struct {
> ```

Quick comment about the purpose of the struct would be helpful.
___
*[pkg/sql/traceanalyzer.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKba7jgBiWCBvBFz7iA:-MKba7jgBiWCBvBFz7iB:b-20bgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer.go#L43)):*
> ```Go
> 
> func (a *TraceAnalyzer) populateIDMaps(flows map[roachpb.NodeID]*execinfrapb.FlowSpec) {
> 	for k := range a.processorIDMap {
> ```

nit: these two `for` loops could be unified into one similar to what you have in `extractStatsByID`.

Also, would do you think about actually merging `populateIDMaps` and `extractStatsByID` methods into one and define a separate helper method that deletes all entries from all 4 maps and calling this new helper method in the beginning of the merged method?
___
*[pkg/sql/traceanalyzer_test.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbbXJz3SwWTHkLH_6f:-MKbbXJz3SwWTHkLH_6g:b-ofvkuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer_test.go#L54)):*
> ```Go
> 							return nil
> 						}
> 						fmt.Println(stmt)
> ```

Is this a leftover?
___
*[pkg/sql/traceanalyzer_test.go, line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbc4HV6Q97cm9b17OU:-MKbc4HV6Q97cm9b17OV:bdrihe0) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer_test.go#L143)):*
> ```Go
> 			{
> 				analyzer:      colexecTraceAnalyzer,
> 				expectedBytes: 1872,
> ```

Just curious: do you know why there is such a big difference in two numbers? Maybe add a quick comment if so.
___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbcl0U9cLocFGvgkEc:-MKbcl0VD-TWLw8IK89S:b5c2799) ([raw file](https://github.com/cockroachdb/cockroach/blob/330b0be83d7bbdd5c2d9d99f230c45708651bb91/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L59)):*
> ```
> SELECT url FROM [EXPLAIN ANALYZE SELECT * FROM kv JOIN kw ON kv.k = kw.k]
> ----
> https://cockroachdb.github.io/distsqlplan/decode.html#eJzMWNFu2zYUfd9XEPepXelKpKQ0FlDA2ZYBLlK7i_OwrfCDYnG2YEfSKCpuEOTfB1kLEjs2RZmipbdYJnXOuT73hLyPkP27Ah8ml1eXv96gnK_Q79fjr-j75Z_fri6GI3Qxurj66-9L9O634eRm8sfVe_T_0p_Lhct79GU8HKHlGo1HaHn_cYk-o-X643IKGOIkZKPgjmXgfwcCGChgcACDCxg8mGJIeTJjWZbwYsnjZsMw_AG-jSGK01wUj6cYZgln4D-CiMSKgQ83we2KXbMgZNyyAUPIRBCtNjDL-0HKo7uAPwCGSRrEmY96VgF8G4jZgmUoyUWaCx8VG0Werl49KjgOx0hEd8xHdlZsehAsQ5wFoY9oH_0CGHiyfn5CYPqEodxdKnhmePuAFkG22OY2KNZPMWQimDPwyRM-TvLZjuT1W8kW6ahoelD0y3sSHjLOwt33fCiAlVbtqd9XxufsSxLFjFtkxzMr9o94NyAf3n_m0XxR_gkYxkV1BgQPKB44eOCql5P9YLNcREn8UtXtmr3Uw1GoRx7v07pX5ijpJalFvG2BCrS9erTdLdpE3bukul0t2rOcLnqXHCv6k0LDdlU0PSj6hA1LutqwVN0EVMH5Ts-qwbot59cQfa7g_K6KpgdFn9D5tKvOd9RN4Cg43-1ZXhdNQI4V3VdwfldF04OiT-h8p6vOd9VN4Co43-t10QLkWMlete97FkFBHCKCErFgvIvy6UH5J-wAt6sdUHFXvWZZmsQZ27mu7H-zXZSBhXNWFjdLcj5j33gy28CUH8ebfZvTZ8gyUX5Lyw_D-PmrTARC9bpjH5Rfh8KZLoXXhq6J3dfBrmgm-3xvN9WhR7R_Ho3aEM8oOKnhS2rGl3IKZ7oUNOT3dbCb8KWcHtH-eTRqs-PLpsHpLrj9GtzZ7ondzY50M9m2lG3G1a7ZwJcWT46tnfT68rUCX1O-4TSvANdK8yYixTMb-NLayLG1k17Fl3IKWoGvKd9wmleAa6V5E748qxP4TdJzqK1A71OLcS7HNntwl2Nr57hKy8op6MV5E86t4Gf2_H7eYpzLsc0e3OXY2jmu4ks5Bb04b8KXFfzMnt_73Y5zYngeI79YGZzENHLrNDit0S2O_uldaSTR7kzG8FBGB1wr1Rsxp8GRjW5x9I_wSuZsdTBDak1mWkj2N8Of5pK9Ef_K-Zk9zVeAG052ObjhuUwVunakK7VunalgG-FaZ3LYeLLUmRmeGNz0qL3W1NDU_5U3I6KWk3369NN_AQAA__-DJN1Z
> ```

nit: I see that we're not showing the number of bytes sent if that number is 0 although we do show some other fields, what do you think about showing zero bytes too? I'm not sure which way I'd prefer, but I wanted to point it out.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cathymw, @dhartunian, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 608 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MK_w_Zp9b6SusBMWwLn:-MKeAzNF6k9QsNrEtmzJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/distsql_physical_planner.go#L608)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: `s/this physical/the physical/` since it's not clear what "this" refers to.
</blockquote></details>

Done.
___
*[pkg/sql/traceanalyzer.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbSwo84ESkg9n9oEOI:-MKeBDh7EeDydU970Ws7:bxyeego) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer.go#L22)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: we should use our `cockroachdb/errors` :)
</blockquote></details>

Done. Thanks. Didn't understand why `go.mod` was changing.
___
*[pkg/sql/traceanalyzer.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKb_gaE8aQVPkLv82VJ:-MKeCRAZCIRR7geh1gkh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer.go#L27)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Quick comment about the purpose of the struct would be helpful.
</blockquote></details>

Done.
___
*[pkg/sql/traceanalyzer.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKba7jgBiWCBvBFz7iA:-MKeCZk3DxDR6eVNWJfH:b2hsien) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer.go#L43)):*
> these two for loops could be unified into one similar to what you have in extractStatsByID.

I can't because the map elements are of different types (`roachpb.NodeID` vs `[2]roachpb.NodeID`)

> Also, would do you think about actually merging populateIDMaps and extractStatsByID methods into one and define a separate helper method that deletes all entries from all 4 maps and calling this new helper method in the beginning of the merged method?

`populateIDMaps` and `extractStatsByID` feel like logically distinct operations so it seems nicer to separate them (I could have kept all that code in this `Reset` method). Is there a reason you'd like to see all of this in a single code block minus the map-resetting?
___
*[pkg/sql/traceanalyzer_test.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbbXJz3SwWTHkLH_6f:-MKeFI_WCTcDZ4KMj5Mw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer_test.go#L54)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Is this a leftover?
</blockquote></details>

Done.
___
*[pkg/sql/traceanalyzer_test.go, line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbc4HV6Q97cm9b17OU:-MKeHcrf2eQplaN18h1b:b-pctlc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer_test.go#L143)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Just curious: do you know why there is such a big difference in two numbers? Maybe add a quick comment if so.
</blockquote></details>

I think it's just the overhead from arrow serialization as well as the fact that we try to encode ints in the least number of bytes possible in the rowexec case. Added a comment.
___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbcl0U9cLocFGvgkEc:-MKeIOuLEowYShx-Lj65:b-pdh250) ([raw file](https://github.com/cockroachdb/cockroach/blob/330b0be83d7bbdd5c2d9d99f230c45708651bb91/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L59)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I see that we're not showing the number of bytes sent if that number is 0 although we do show some other fields, what do you think about showing zero bytes too? I'm not sure which way I'd prefer, but I wanted to point it out.
</blockquote></details>

I'm not sure. @cathymw what do you think?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great stuff!

Until we have always-on tracing, we could collect this any time we collect the plan (see `planTop.savePlanForStats`)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-:-MKf8bws0JAtXPjKL3QJ:b-1fv0na)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, @dhartunian, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 971 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf14zx4CO8KnhWfzZg:-MKf14zx4CO8KnhWfzZh:b-j4nfh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/conn_executor_exec.go#L971)):*
> ```Go
> 				planCtx.saveFlows = fn
> 			}
> 		} else {
> ```

[nit] can do `else if planner.collectBundle` here and remove the outer if.

I assume we don't care about support bundles when TestingSaveFlows is set.
___
*[pkg/sql/conn_executor_exec.go, line 399 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf8COVBpB_BLk9gm8I:-MKf8COVBpB_BLk9gm8J:b-qyv9v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d24f23466ee788573bb32554a64fb4d91c5c6ca/pkg/sql/conn_executor_exec.go#L399)):*
> ```Go
> 				if err != nil && log.V(1) {
> 					log.Infof(ctx, "error calculating network bytes sent for stmt %s: %v", stmt, err)
> 					continue
> ```

We should `continue` regardless of `log.V` (maybe we can do log.VInfof() instead of checking, that has the benefit that it will always show up in traces)
___
*[pkg/sql/exec_util.go, line 853 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf2-sG1sd1mG16GwbA:-MKf2-sG1sd1mG16GwbB:b-ipz3tv) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/exec_util.go#L853)):*
> ```Go
> 	// TestingSaveFlows, if set, will be called with the given stmt. The resulting
> 	// function will be called with the physical plan of that statement's main
> 	// query (i.e. no subqueries).
> ```

Mention that `nil` can be returned (though maybe returning a no-op func would be cleaner).
___
*[pkg/sql/plan.go, line 270 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf6Hmq6Hl6eAm6mDcx:-MKf6Hmq6Hl6eAm6mDcy:b-l0i12k) ([raw file](https://github.com/cockroachdb/cockroach/blob/246ab61dd57c94bfd6abd2576bf4b18bcd050910/pkg/sql/plan.go#L270)):*
> ```Go
> var _ planNodeSpooled = &spoolNode{}
> 
> type queryType int
> ```

The name is very general, how about "planComponentType"? (these things are grouped in the `planComponents` struct). We could also move the defs next to that struct.
___
*[pkg/sql/plan.go, line 288 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf78687sLGXBx65lm9:-MKf78687sLGXBx65lmA:b-e9h6xw) ([raw file](https://github.com/cockroachdb/cockroach/blob/246ab61dd57c94bfd6abd2576bf4b18bcd050910/pkg/sql/plan.go#L288)):*
> ```Go
> 		return "postquery"
> 	default:
> 		return "<unknown>"
> ```

[nit] maybe remove the brackets, since this makes it into a filename
___
*[pkg/sql/traceanalyzer.go, line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf2dLTE0QpOIsfgWaM:-MKf2dLTE0QpOIsfgWaN:byhw20f) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L33)):*
> ```Go
> //     analyzer.Reset(flows, trace)
> //     bytesGroupedByNode, err := analyzer.GetNetworkBytesSent()
> type TraceAnalyzer struct {
> ```

This feels generic enough that it doesn't need to live in `sql`. Is there an existing package we can use? Or we could make a new `sql/execstats` package.
___
*[pkg/sql/traceanalyzer.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf3L2D7Liczyakxz1z:-MKf3L2D7Liczyakxz2-:be249up) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L35)):*
> ```Go
> type TraceAnalyzer struct {
> 	// processorIDMap maps a processor ID to a node it was planned on.
> 	processorIDMap map[int]roachpb.NodeID
> ```

We should make separate types for the processor ID and stream ID (or use the ones from the specs), to make things more evident
___
*[pkg/sql/traceanalyzer.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf38tUEKuYCOBZ723_:-MKf38tUEKuYCOBZ723a:b-8qx1n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L38)):*
> ```Go
> 	// streamIDMap maps a stream ID to a given pair of nodes. The first node in
> 	// the array is the origin node, the second is the destination node.
> 	streamIDMap map[int][2]roachpb.NodeID
> ```

I know it's a bit of a pain, but making a struct with field names makes the code easier to read
___
*[pkg/sql/traceanalyzer.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf41W77JLv2w1T_RKC:-MKf41W77JLv2w1T_RKD:bl61ydb) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L42)):*
> ```Go
> 	// processorStats maps a processor ID to the stats associated with this
> 	// processor extracted from a trace.
> 	processorStats map[int]execinfrapb.DistSQLSpanStats
> ```

It feels a little strange that we have two maps with the same key, can we just put the NodeID and the stats in a struct and use a single map?
___
*[pkg/sql/traceanalyzer.go, line 45 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf48ju-uVMrvNYxFRG:-MKf48ju-uVMrvNYxFRH:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L45)):*
> ```Go
> 	// streamStats maps a stream ID to the stats associated with this stream
> 	// extracted from a trace.
> 	streamStats map[int]execinfrapb.DistSQLSpanStats
> ```

ditto
___
*[pkg/sql/traceanalyzer_test.go, line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf4ROyEU1CIGWFJlDz:-MKf4ROyEU1CIGWFJlE-:b-yfvslt) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer_test.go#L34)):*
> ```Go
> )
> 
> func TestTraceAnalyzer(t *testing.T) {
> ```

[nit] It would be nice to give a high level overview of what the test does
___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 54 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf7SOX3qkgiAaYY8Ob:-MKf7SOX3qkgiAaYY8Oc:b-wg3m4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L54)):*
> ```
> SELECT url FROM [EXPLAIN ANALYZE SELECT count(*) FROM kv]
> ----
> https://cockroachdb.github.io/distsqlplan/decode.html#eJzMld1um0AQhe_7FKu5squ1MD92bK7ipq6E5EBqHPUnQhGBkYuCWbq7pIksv3sFpAq2HKCqGnPJsIc5h29gtiB-xmCCO1_ML1Yk4zH5tHQuyc3869ViZtlkZs8W377PSe-j5a7cz4s-eT4asCyRvff98vz9gwcUEhai7W9QgHkDKlDQgIIOFAygMAKPQspZgEIwnh_ZFgIrfARzSCFK0kzmZY9CwDiCuQUZyRjBhJV_F-MS_RC5MgQKIUo_ios29w_nKY82Pn8CCm7qJ8IkAyVv7GTSJDZLECjc-TL4gYKwTKZ5OX-IzNK4Usr9Wg6R0QZNMhS56EmiIBz90CTalHwACpz9-lNRwdtRKNXPtoX01wimuqPto83Wa45rXzKujPaTXTjX9up26Xxxe_32EfARg0xGLHlJ8ppP7VWfL_ayhPEQOYZ73rxdfRL1AJJ7fXlr2aveufp_kuh7SdT2w6Q2D5OiDRS9O-PUEK4CYXzScdLaQ9BaQNAHitEdCA3hKhDOTgpBbw9BbwHBGBT_qI5AaAhXgTA5KQSjPQSjBYTRoDsIGqJVEEw7s9uO-FyiSFki8GDHHX_yMN99GK6xXJSCZTzAK86Cok156RS6ohCikOVdtbywkvJWbrAqVmvF2p5YPRRr9Z0bWuu1aqNebPyN7-ItFi_038dWHWlH53bf3qjW3rg-27jb2c5q7U3qs026nW1aP9HDho-p_lN8-3Te7t3vAAAA__8nPFQM
> ```

What caused this change? Why weren't we showing the bytes read before?


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, @dhartunian, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbcl0U9cLocFGvgkEc:-MKg5q1ZAjKTuyB0Nx32:bl36yp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/330b0be83d7bbdd5c2d9d99f230c45708651bb91/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L59)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm not sure. @cathymw what do you think?
</blockquote></details>

We also don't show rows read if it's 0. I don't have a strong preference either. It might be nice to be consistent. I'm not sure if it would be confusing for people to not see all the stats, especially if other stats still get displayed when they have zero values. I don't know if it would look too crowded if the diagrams get very large and complicated though.
___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 54 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf7SOX3qkgiAaYY8Ob:-MKg13Me41spw10cLHQi:b-ve232m) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L54)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What caused this change? Why weren't we showing the bytes read before?
</blockquote></details>

I think it's because when bytesRead == 0, they aren't displayed. And then, prior to these changes, for tests we would set bytes read to zero due to non-determinism. But now, Alfonso fixed the non-determinism, so the value no longer gets zeroed out and shows up on the diagram now


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
TFTR. Addressed all comments but I'm still working on a test flake so this is not ready to merge yet.

> Until we have always-on tracing, we could collect this any time we collect the plan (see planTop.savePlanForStats)

Do we enable tracing in those cases? It doesn't seem like it but I may be missing something.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-:-MKigi7a9AbqwnVUq7cI:bj3y2ky)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cathymw, @dhartunian, @jordanlewis, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 399 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf8COVBpB_BLk9gm8I:-MKigxbX7RNbMQVnxkIA:bfv3b4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d24f23466ee788573bb32554a64fb4d91c5c6ca/pkg/sql/conn_executor_exec.go#L399)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should `continue` regardless of `log.V` (maybe we can do log.VInfof() instead of checking, that has the benefit that it will always show up in traces)
</blockquote></details>

Oops, good catch.
___
*[pkg/sql/plan.go, line 270 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf6Hmq6Hl6eAm6mDcx:-MKjL4Ta330nKsaPaPOa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/246ab61dd57c94bfd6abd2576bf4b18bcd050910/pkg/sql/plan.go#L270)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The name is very general, how about "planComponentType"? (these things are grouped in the `planComponents` struct). We could also move the defs next to that struct.
</blockquote></details>

Done.
___
*[pkg/sql/traceanalyzer.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKba7jgBiWCBvBFz7iA:-MKjFaSjEffPVmMu0hs9:b-7baiqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/739379b5f73920b0fe6c607a753bd716e8cce767/pkg/sql/traceanalyzer.go#L43)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

> these two for loops could be unified into one similar to what you have in extractStatsByID.

I can't because the map elements are of different types (`roachpb.NodeID` vs `[2]roachpb.NodeID`)

> Also, would do you think about actually merging populateIDMaps and extractStatsByID methods into one and define a separate helper method that deletes all entries from all 4 maps and calling this new helper method in the beginning of the merged method?

`populateIDMaps` and `extractStatsByID` feel like logically distinct operations so it seems nicer to separate them (I could have kept all that code in this `Reset` method). Is there a reason you'd like to see all of this in a single code block minus the map-resetting?
</blockquote></details>

I ended up putting everything together in `Reset`.
___
*[pkg/sql/traceanalyzer.go, line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf2dLTE0QpOIsfgWaM:-MKimtObDDJI8cMXNi-8:b-vg9ngy) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L33)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This feels generic enough that it doesn't need to live in `sql`. Is there an existing package we can use? Or we could make a new `sql/execstats` package.
</blockquote></details>

That was my initial approach, but an import cycle occurs in the test because `serverutils` depends on `sql` which in turn depends on `execstats`. It's unfortunate that a test causes this. Do you see any way to break the cycle? Another option is to keep the test in `sql` but that seems unfortunate.
___
*[pkg/sql/traceanalyzer.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf41W77JLv2w1T_RKC:-MKioe1l6SfaaSxjBrrj:b-yh16l6) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L42)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It feels a little strange that we have two maps with the same key, can we just put the NodeID and the stats in a struct and use a single map?
</blockquote></details>

Done. Good call.
___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKbcl0U9cLocFGvgkEc:-MKjHzEFBTZuPRIEOmv9:b-mgi2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/330b0be83d7bbdd5c2d9d99f230c45708651bb91/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L59)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

We also don't show rows read if it's 0. I don't have a strong preference either. It might be nice to be consistent. I'm not sure if it would be confusing for people to not see all the stats, especially if other stats still get displayed when they have zero values. I don't know if it would look too crowded if the diagrams get very large and complicated though.
</blockquote></details>

@jordanlewis was surprised by this behavior as well. I'm happy to change it but will leave it for another PR.
___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 54 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf7SOX3qkgiAaYY8Ob:-MKjHeWh2LCR2qPa8AfV:b-hzxx6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L54)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

I think it's because when bytesRead == 0, they aren't displayed. And then, prior to these changes, for tests we would set bytes read to zero due to non-determinism. But now, Alfonso fixed the non-determinism, so the value no longer gets zeroed out and shows up on the diagram now
</blockquote></details>

That's right. The non-determinism was due to including metadata size in bytes sent. The commit changes the bytes sent calculation to only include the row bytes in tests (i.e. when the `DeterministicStats` flag is set) for determinism.


<!-- Sent from Reviewable.io -->

asubiotto commented:
I think I figured the test issue out. It seems `FlowSpecs` are not safe for use after `saveFlows` is called as we currently do in the test since they're released to the sync pool. I have to step back from this PR for today but I'll have to think about what the best way forward should be. cc @yuzefovich if you've got any suggestions

yuzefovich commented:
Oh yeah, we do pool those objects. I guess we'll have to release the flow specs back to the pool later, after the trace has been analyzed.

dhartunian reviewed (commented):
> Also, I'm curious why we're showing a decimal point in the number of "rows read" in the image, that looks not pretty, cc @dhartunian

@yuzefovich unfortunately the stats table component applies a uniform number formatting to all rows. Once the statements page is embedded into CC, we'll devote more time and attention to these visual nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-:-MKk2Jou5FzyCkUzQ81u:b-783tml)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @cathymw, @dhartunian, @jordanlewis, @RaduBerinde, and @yuzefovich)

___
*[pkg/ui/src/views/statements/statementDetails.tsx, line 508 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKk1ZBD9S3WXLNh_P10:-MKk1ZBD9S3WXLNh_P11:bnv8tq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91e5e8a2db9d1a62a715d31ea44596d9ae1dbf3/pkg/ui/src/views/statements/statementDetails.tsx#L508)):*
> ```XML
>                 },
>               ].filter(function (r) {
>                 console.log("filtering", r.value);
> ```

nit: assume this will be removed before merge.
___
*[pkg/ui/src/views/statements/statementDetails.tsx, line 511 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKk1OR5BUCZ63ZbtGrJ:-MKk1OR5BUCZ63ZbtGrK:b1iad8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91e5e8a2db9d1a62a715d31ea44596d9ae1dbf3/pkg/ui/src/views/statements/statementDetails.tsx#L511)):*
> ```XML
>                 if (r.name === "Network Bytes Sent") {
>                   if (r.value.mean === 0) {
>                     // Omit if empty.
> ```

Curious if there's any difference in how the `rows_read` and `bytes_read` fields are generated and this one, why should we omit showing this when the number is zero? 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> Do we enable tracing in those cases? It doesn't seem like it but I may be missing something.

No, we would need to do that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-:-MKk6mdzCUq9tlef41nI:bo2rgh3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/traceanalyzer.go, line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf2dLTE0QpOIsfgWaM:-MKk7MerDsn1DJiAppET:b8ct2ex) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L33)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

That was my initial approach, but an import cycle occurs in the test because `serverutils` depends on `sql` which in turn depends on `execstats`. It's unfortunate that a test causes this. Do you see any way to break the cycle? Another option is to keep the test in `sql` but that seems unfortunate.
</blockquote></details>

`serverutils` does not depend on `sql`, it exists exactly to allow tests to create test servers and clusters without directly depending on `server` and `sql`. Plus since you are only testing external interfaces, the test would be in a `something_test` package where it can depend on anything.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/traceanalyzer_test.go, line 14 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKm_PPH5LHpS7KHji77:-MKm_PPH5LHpS7KHji78:b-gssig9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7ca0159c10366e067ae0fa5d70446c090bd221/pkg/sql/traceanalyzer_test.go#L14)):*
> ```Go
> 
> import (
> 	context "context"
> ```

[nit] redundant alias
___
*[pkg/sql/traceanalyzer_test.go, line 152 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKm_hPu3wKYKT3qRMIi:-MKm_hPu3wKYKT3qRMIj:bm77a6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7ca0159c10366e067ae0fa5d70446c090bd221/pkg/sql/traceanalyzer_test.go#L152)):*
> ```Go
> 			{
> 				analyzer: colexecTraceAnalyzer,
> 				// Note that the expectedBytes in the colexec case is larger than the
> ```

It's strange that we'll need to update these tests whenever some over the wire format changes. Maybe we should just check that the value is in a reasonable range (e.g. 32 to 4096)


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> No, we would need to do that.

I like the idea but isn't the overhead concerning? Will leave this for later in any case

I changed the`saveFlow` semantics to disallow saving flows past the function's lifetime since we would otherwise need to defer releases up in conn executor space, which seems like a big no-no. The result is that we create analyzers in `saveFlow`, have them transform the flow into the maps they can use, and attach those analyzers to a plan that are then reused by the conn executor.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-:-MKoIF7x9wmXjtMBJGKn:biqyocg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @jordanlewis, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/traceanalyzer.go, line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf2dLTE0QpOIsfgWaM:-MKoGeth3K7F5rPGisvM:b-8v7xwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L33)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`serverutils` does not depend on `sql`, it exists exactly to allow tests to create test servers and clusters without directly depending on `server` and `sql`. Plus since you are only testing external interfaces, the test would be in a `something_test` package where it can depend on anything.
</blockquote></details>

Not directly, but to use the testcluster you need to initialize the factory in `TestMain`:
```
func TestMain(m *testing.M) {
	serverutils.InitTestClusterFactory(testcluster.TestClusterFactory)
}
```
This in turn depends on `server` and `sql`:
```
package github.com/cockroachdb/cockroach/pkg/sql/execstats (test)
	imports github.com/cockroachdb/cockroach/pkg/testutils/testcluster
	imports github.com/cockroachdb/cockroach/pkg/server
	imports github.com/cockroachdb/cockroach/pkg/sql
	imports github.com/cockroachdb/cockroach/pkg/sql/execstats: import cycle not allowed in test
```

Am I missing something? This is with the test already in the `*_test` package.
___
*[pkg/ui/src/views/statements/statementDetails.tsx, line 511 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKk1OR5BUCZ63ZbtGrJ:-MKnkTV51J7U7Ech9MvB:b-33m2xv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91e5e8a2db9d1a62a715d31ea44596d9ae1dbf3/pkg/ui/src/views/statements/statementDetails.tsx#L511)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Curious if there's any difference in how the `rows_read` and `bytes_read` fields are generated and this one, why should we omit showing this when the number is zero? 
</blockquote></details>

There is a difference. The network bytes sent is only collected when a statement diagnostics bundle is requested because of the current overhead of collecting that information. The idea is that by the end of this release, we will get to a point where we can display this stat unconditionally. Until then, it would be misleading to display "0" for this stat if it hasn't been collected.


<!-- Sent from Reviewable.io -->

asubiotto commented:
I'm still on the fence re adding the total network bytes sent for the query. Is that data useful? Would it be better to send back some 2D slice where each cell is the bytes sent from node i -> node j? We could possibly enhance such a data structure in the future with network latency and number of messages sent. We could potentially just deprecate the network bytes sent but it's probably worth thinking through what the best way forward here is.

One problem with the per-node approach is that physical plans can change and nodes might not be part of a physical plan any longer. Is simply resetting the previous

RaduBerinde commented:
___
*[pkg/sql/traceanalyzer.go, line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-MKf2dLTE0QpOIsfgWaM:-MKpjXWDEJI82qvO54vw:b-36lsuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ba3bd446716f8ba4bcce3bb0ff7785aa8a9f9f/pkg/sql/traceanalyzer.go#L33)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Not directly, but to use the testcluster you need to initialize the factory in `TestMain`:
```
func TestMain(m *testing.M) {
	serverutils.InitTestClusterFactory(testcluster.TestClusterFactory)
}
```
This in turn depends on `server` and `sql`:
```
package github.com/cockroachdb/cockroach/pkg/sql/execstats (test)
	imports github.com/cockroachdb/cockroach/pkg/testutils/testcluster
	imports github.com/cockroachdb/cockroach/pkg/server
	imports github.com/cockroachdb/cockroach/pkg/sql
	imports github.com/cockroachdb/cockroach/pkg/sql/execstats: import cycle not allowed in test
```

Am I missing something? This is with the test already in the `*_test` package.

</blockquote></details>

The trick is that TestMain should be in `_test` package (which can depend on anything).


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> from admin ui perspective, thanks for answering my question

>I'm still on the fence re adding the total network bytes sent for the query. Is that data useful? Would it be better to send back some 2D slice where each cell is the bytes sent from node i -> node j? We could possibly enhance such a data structure in the future with network latency and number of messages sent. We could potentially just deprecate the network bytes sent but it's probably worth thinking through what the best way forward here is.

My take is to keep the data simpler in the short run. We already have more data in the Admin UI than we can properly visualize and make actionable so I hesitate to add extra detail unless we have a clear way to signal when it matters to someone. 

> Reviewed 3 of 13 files at r1, 1 of 18 files at r7, 19 of 19 files at r10, 6 of 6 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55969#-:-MKqE6h02CUnvSocwpok:b-9x1lui)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @dhartunian, @jordanlewis, and @yuzefovich)



<!-- Sent from Reviewable.io -->

asubiotto commented:
Thanks @RaduBerinde, that worked. Put the `TraceAnalyzer` in its own `execstats` package.

asubiotto commented:
I think I've convinced myself that it's useful to have this statement-level statistic in any case. It keeps things simple as @dhartunian mentions, and is not incompatible with statistics in the context of a physical plan which we can return in addition to statement-level stats. It probably makes sense attach these to an EXPLAIN ANALYZE representation as extra metadata. Will bors on green.

Commit:
sql/ui: collect and display network bytes sent per statement

This commit uses a TraceAnalyzer when statement diagnostics are enabled to
populate BytesSentOverNetwork on a per-statement basis. This data is displayed
in the corresponding statement's page if collected.

This work is a precursor to always-on tracing. Once always-on tracing is
enabled, these types of statement statistics calculated from the trace will
always be displayed in the admin UI.

Release note (admin ui change): If statement diagnostics are enabled for a
statement, the bytes sent over the network will be shown in the statement's
page.

Commit:
sql: add TraceAnalyzer

This commit adds TraceAnalyzer, which calculates general stats from a physical
plan and an accompanying statement trace. This commit adds an example
calculation of the number of netowrk bytes sent by each node in the plan.

Additionally, this commit fixes non-determinism in outbox network statistics
caused by variable metadata in order to make testing more useful.

Release note: None

Commit:
sql: unify saveFlows code and add query types to diagrams and physical plans

This commit factors out some repeated code and adds a query type to flow
diagrams, which was a TODO when collecting statement bundles in order to be
able to differentiate between main queries, postqueries, and subqueries.

Release note: None

asubiotto commented:
Latest CI failure seems like an infra issue. TFTRs

bors r=RaduBerinde,dhartunian
